service/storage_proxy.cc

```diff
@@ -4136,6 +4136,32 @@ storage_proxy::query_partition_key_range_concurrent(storage_proxy::clock_type::t
                 break;
             }
 
+            // Checks if 2 range bounds are contiguous, permits false negatives
+            auto are_contiguous = [schema](const range_bound<dht::ring_position>& first_range_end,
+                                           const range_bound<dht::ring_position>& second_range_start) {
+                // Implementing a proper contiguity check is hard, because it requires
+                // is_successor(range_bound<dht::ring_position> a, range_bound<dht::ring_position> b)
+                // relation to be defined. It is needed for intervals for which their possibly adjacent
+                // bounds are either both exclusive or inclusive.
+                // For example: is_adjacent([a, b], [c, d]) requires checking is_successor(b, c).
+                // Defining a successor relationship for dht::ring_position is hard, because
+                // dht::ring_position can possibly contain partition key.
+                // Luckily, a full contiguity check here is not needed.
+                // Ranges that we want to merge here are formed by dividing a bigger ranges using
+                // query_ranges_to_vnodes_generator. By knowing query_ranges_to_vnodes_generator internals,
```

Comment:
I think this statement only tells half the story. Yes, the **usual** case is that a range is split into ranges with left-closed ranges [a, b), [b, c).  However, I think the whole point of this fix was to support new code (COUNT, Alternator TTL, etc.) which sends a `partition_ranges` vector which has multiple unrelated ranges. So you can't make assumptions on these ranges. However, I think your check is safe if we assume that it's fine that we don't merge in some cases (like [a, b] [b+1, c]) because not merging is not a correctness problem. Maybe this is the "false negative" you mentioned int he above comment but I didn't understand.